core/trino-main/src/main/java/io/trino/operator/output/BytePositionsAppender.java

```diff
@@ -14,31 +14,239 @@
+    private byte[] values = new byte[0];
+
+    private long retainedSizeInBytes;
+
+    public BytePositionsAppender(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this.blockBuilderStatus = blockBuilderStatus;
+        this.initialEntryCount = max(expectedEntries, 1);
+
+        updateDataSize();
+    }
+
+    @Override
+    public void appendTo(IntArrayList positions, Block block)
+    {
+        int[] positionArray = positions.elements();
+        int newPositionCount = positions.size();
+        ensureCapacity(positionCount + newPositionCount);
+        boolean hasNullValue = false;
+        boolean hasNonNullValue = false;
+
+        if (block.mayHaveNull()) {
+            for (int i = 0; i < newPositionCount; i++) {
+                int position = positionArray[i];
+                boolean isNull = block.isNull(position);
+                int positionIndex = positionCount + i;
+                if (isNull) {
+                    valueIsNull[positionIndex] = true;
+                    hasNullValue = true;
+                }
+                else {
+                    values[positionIndex] = block.getByte(position, 0);
+                    hasNonNullValue = true;
+                }
+            }
+            this.positionCount += newPositionCount;
+            this.hasNonNullValue |= hasNonNullValue;
+            this.hasNullValue |= hasNullValue;
+        }
+        else {
+            for (int i = 0; i < newPositionCount; i++) {
+                int position = positionArray[i];
+                values[positionCount + i] = block.getByte(position, 0);
+            }
+            positionCount += newPositionCount;
+            this.hasNonNullValue = true;
+        }
+
+        if (blockBuilderStatus != null) {
+            blockBuilderStatus.addBytes(ByteArrayBlock.SIZE_IN_BYTES_PER_POSITION * newPositionCount);
+        }
+    }
+
     @Override
-    public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+    public Block build()
+    {
+        if (!hasNonNullValue) {
+            return new RunLengthEncodedBlock(NULL_VALUE_BLOCK, positionCount);
+        }
+        return new ByteArrayBlock(positionCount, hasNullValue ? Optional.of(valueIsNull) : Optional.empty(), values);
+    }
+
+    @Override
+    public void appendDictionary(IntArrayList positions, DictionaryBlock block)
     {
         int[] positionArray = positions.elements();
+        int newPositionCount = positions.size();
+        ensureCapacity(positionCount + newPositionCount);
         if (block.mayHaveNull()) {
-            for (int i = 0; i < positions.size(); i++) {
+            for (int i = 0; i < newPositionCount; i++) {
                 int position = positionArray[i];
                 if (block.isNull(position)) {
-                    blockBuilder.appendNull();
+                    valueIsNull[positionCount] = true;
+                    hasNullValue = true;
```

Comment:
Why no local args to hold temporary `hasNullValue` and `hasNonNullValue`. Did you benchmark if making them local actually helps?